### PR TITLE
Add Select2 AJAX for Terms

### PR DESCRIPTION
**Pull request by bordoni**
_Ref:_ [C#77125](http://central.tri.be/issues/77125)
---

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $page is never used | [Dropdown.php:15](https://github.com/moderntribe/tribe-common/blob/7dcaad4ba4324f230123ca5615f175fce298d8ae/src/Tribe/Ajax/Dropdown.php#L15)
:warning: | The method parameter $source is never used | [Dropdown.php:15](https://github.com/moderntribe/tribe-common/blob/7dcaad4ba4324f230123ca5615f175fce298d8ae/src/Tribe/Ajax/Dropdown.php#L15)
:warning: | The method parameter $arguments is never used | [Dropdown.php:152](https://github.com/moderntribe/tribe-common/blob/7dcaad4ba4324f230123ca5615f175fce298d8ae/src/Tribe/Ajax/Dropdown.php#L152)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Ajax/Dropdown.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			}
+		}
+
+		$data['results'] = $results;
+		$data['taxonomies'] = get_taxonomies();
+
+		return $data;
+	}
+
+	public function sort_terms_hierarchicaly( &$terms, &$into, $parent = 0 ) {
```
Comment (barryhughes) [inline src/Tribe/Ajax/Dropdown.php]: Typo: hierarchically

File: src/resources/js/dropdowns.js [hunk 2]
```diff
@@ -298,20 +298,51 @@ var tribe_dropdowns = tribe_dropdowns || {};
...
+						return { results: [] };
+					}
+
+					if ( ! $.isPlainObject( response.data ) || 'undefined' === typeof response.data.results ) {
+						console.error( 'We received a malformed results array, could not complete the Select2 Search.' );
+						return { results: [] };
+					}
+
+					if ( ! response.success ) {
+						if ( 'string' === jQuery.type( response.data.message ) ) {
```
Comment (barryhughes) [inline src/resources/js/dropdowns.js]: Simplify `jQuery` to `$`?

File: src/Tribe/Ajax/Dropdown.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ * Handles common AJAX operations.
+ */
+class Tribe__Ajax__Dropdown {
+
+	public function hook() {
+		add_action( 'wp_ajax_tribe_dropdown', array( $this, 'route' ) );
+		add_action( 'wp_ajax_nopriv_tribe_dropdown', array( $this, 'route' ) );
+	}
+
+	public function search_terms( $search, $page, $args, $source ) {
```
Comment (borkweb) [inline src/Tribe/Ajax/Dropdown.php]: PHPDoc comment?

File: src/Tribe/Ajax/Dropdown.php (hunk 1 above)
Comment (borkweb) [inline src/Tribe/Ajax/Dropdown.php]: Also: PHPDoc comment

Comment (borkweb) [inline src/Tribe/Ajax/Dropdown.php]: Rinse and repeat for the other methods.

File: src/Tribe/Ajax/Dropdown.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+		$arguments = wp_parse_args( $params, $defaults );
+
+		// Return Object just for the sake of making it simpler to read
+		return (object) $arguments;
+	}
+
+	public function route() {
+		// Push all POST params into a Default set of data
+		$args = $this->parse_params( $_POST );
```
Comment (borkweb) [inline src/Tribe/Ajax/Dropdown.php]: Should we do:

```php
$args = $this->parse_params( empty( $_POST ) ? array() : $_POST );
```

File: src/resources/css/tribe-common-admin.css [hunk 5]
```diff
@@ -1002,6 +1010,28 @@ a.tribe-rating-link {
 	-webkit-transition-property: border,color;
 	transition-property: border,color;
 }
+.select2-results .select2-selected {
+	display: block;
+}
+.select2-results .select2-selected > .select2-result-label {
+	background-color: #efefef;
+	cursor: default;
+	color: #a1a1a1;
```
Comment (borkweb) [inline src/resources/css/tribe-common-admin.css]: Alphabetize

File: src/resources/postcss/tribe-common-admin.pcss [hunk 6]
```diff
@@ -1200,7 +1207,36 @@ a.tribe-rating-link {
...
 
 .select2-results {
+
+	.select2-selected {
+		display: block;
+
+		> .select2-result-label {
+			background-color: #efefef;
+			cursor: default;
+			color: #a1a1a1;
```
Comment (borkweb) [inline src/resources/postcss/tribe-common-admin.pcss]: Alphabetize

File: src/resources/postcss/tribe-common-admin.pcss [hunk 7]
```diff
@@ -1200,7 +1207,36 @@ a.tribe-rating-link {
...
+			display: block;
+
+			.select2-result {
+				display: block;
+			}
+
+			> .select2-result-label {
+				background-color: #efefef;
+				cursor: default;
+				color: #a1a1a1;
```
Comment (borkweb) [inline src/resources/postcss/tribe-common-admin.pcss]: Alphabetize

File: src/resources/js/dropdowns.js [hunk 8]
```diff
@@ -298,20 +298,51 @@ var tribe_dropdowns = tribe_dropdowns || {};
...
 			// By default only send the source
 			args.ajax.data = function( search, page ) {
 				return {
-					// @todo remove aggregator reference
-					action: 'tribe_aggregator_dropdown_' + source,
+					action: 'tribe_dropdown',
+					source: source,
+					search: search,
+					page: page,
+					args: $select.data( 'source-args' ),
```
Comment (borkweb) [inline src/resources/js/dropdowns.js]: Remove the trailing comma

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $page is never used | [Dropdown.php:15](https://github.com/moderntribe/tribe-common/blob/cf6c0cfbcae3da94d026df67456457a067ff4a0c/src/Tribe/Ajax/Dropdown.php#L15)
:warning: | The method parameter $source is never used | [Dropdown.php:15](https://github.com/moderntribe/tribe-common/blob/cf6c0cfbcae3da94d026df67456457a067ff4a0c/src/Tribe/Ajax/Dropdown.php#L15)
:warning: | The method parameter $arguments is never used | [Dropdown.php:152](https://github.com/moderntribe/tribe-common/blob/cf6c0cfbcae3da94d026df67456457a067ff4a0c/src/Tribe/Ajax/Dropdown.php#L152)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $page is never used | [Dropdown.php:34](https://github.com/moderntribe/tribe-common/blob/6cdc49364a2941ec08a2ed29fe105fd38453f49c/src/Tribe/Ajax/Dropdown.php#L34)
:warning: | The method parameter $source is never used | [Dropdown.php:34](https://github.com/moderntribe/tribe-common/blob/6cdc49364a2941ec08a2ed29fe105fd38453f49c/src/Tribe/Ajax/Dropdown.php#L34)
:warning: | The method parameter $arguments is never used | [Dropdown.php:233](https://github.com/moderntribe/tribe-common/blob/6cdc49364a2941ec08a2ed29fe105fd38453f49c/src/Tribe/Ajax/Dropdown.php#L233)

> via jenkins/codesniffer-phpcs

**Comment by borkweb** [comment]
:+1:

**Pull request merged**